File: android/app/src/main/java/app/farmsy/android/core/Contributions.kt
package app.farmsy.android.core

import androidx.annotation.StringRes
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material.icons.filled.Map
import androidx.compose.material.icons.filled.PhotoCamera
import androidx.compose.material.icons.filled.ShoppingBasket
import androidx.compose.material.icons.filled.Verified
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.ui.graphics.vector.ImageVector
import app.farmsy.android.R
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpHeaders
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Badges and the leaderboard: what a person gave the map, as the server counts
// it (migration 069). The app only reads; award_badges runs server-side after
// every report and post, so nothing here can be earned by tapping.
// The Android twin of iOS Contributions.swift.

/// The eight badges, in display order, with their rule in words.
enum class BadgeKind(val wire: String, @StringRes val titleRes: Int, @StringRes val ruleRes: Int, val icon: ImageVector) {
    FIRST_REPORT("first_report", R.string.badge_first_report, R.string.badge_rule_first_report, Icons.Filled.Flag),
    EARLY_BIRD("early_bird", R.string.badge_early_bird, R.string.badge_rule_early_bird, Icons.Filled.WbSunny),
    CONFIRMER("confirmer", R.string.badge_confirmer, R.string.badge_rule_confirmer, Icons.Filled.Verified),
    EXPLORER("explorer", R.string.badge_explorer, R.string.badge_rule_explorer, Icons.Filled.Map),
    SHELF_SCOUT("shelf_scout", R.string.badge_shelf_scout, R.string.badge_rule_shelf_scout, Icons.Filled.ShoppingBasket),
    PHOTOGRAPHER("photographer", R.string.badge_photographer, R.string.badge_rule_photographer, Icons.Filled.PhotoCamera),
    REGULAR("regular", R.string.badge_regular, R.string.badge_rule_regular, Icons.Filled.CalendarMonth),
    COLLECTOR("collector", R.string.badge_collector, R.string.badge_rule_collector, Icons.Filled.Favorite);

    companion object {
        fun fromWire(s: String): BadgeKind? = entries.firstOrNull { it.wire == s }
    }
}

@Serializable
data class ContributionStats(
    val reports: Int = 0,
    val farms: Int = 0,
    val confirmations: Int = 0,
    @SerialName("early_birds") val earlyBirds: Int = 0,
    @SerialName("shelf_reports") val shelfReports: Int = 0,
    val weeks: Int = 0,
    val posts: Int = 0,
    @SerialName("photo_posts") val photoPosts: Int = 0,
    val saved: Int = 0,
    val badges: Int = 0,
)

@Serializable
data class EarnedBadge(val badge: String, @SerialName("earned_at") val earnedAt: String)

@Serializable
data class LeaderRow(
    @SerialName("user_id") val userId: String,
    val name: String,
    val reports: Int = 0,
    val farms: Int = 0,
    val confirmations: Int = 0,
    val badges: Int = 0,
)

object Contributions {
    private val _stats = MutableStateFlow<ContributionStats?>(null)
    val stats: StateFlow<ContributionStats?> = _stats.asStateFlow()
    private val _earned = MutableStateFlow<List<EarnedBadge>>(emptyList())
    val earned: StateFlow<List<EarnedBadge>> = _earned.asStateFlow()
    /// Badges the last refresh awarded — the profile shows a celebration for
    /// these once, then forgets them.
    private val _justAwarded = MutableStateFlow<List<BadgeKind>>(emptyList())
    val justAwarded: StateFlow<List<BadgeKind>> = _justAwarded.asStateFlow()
    private val _leaderboard = MutableStateFlow<List<LeaderRow>>(emptyList())
    val leaderboard: StateFlow<List<LeaderRow>> = _leaderboard.asStateFlow()
    private val _leaderboardMonth = MutableStateFlow<List<LeaderRow>>(emptyList())
    val leaderboardMonth: StateFlow<List<LeaderRow>> = _leaderboardMonth.asStateFlow()

    fun has(kind: BadgeKind): Boolean = _earned.value.any { it.badge == kind.wire }
    fun clearAwarded() { _justAwarded.value = emptyList() }

    suspend fun refreshMine(token: String) {
        val decoded = runCatching {
            val resp = httpClient.get("${Backend.WEB_API}/profile/contributions") {
                header(HttpHeaders.Authorization, "Bearer $token")
            }
            if (resp.status.value != 200) null
            else lenientJson.decodeFromString<MinePayload>(resp.bodyAsText())
        }.getOrNull() ?: return
        _stats.value = decoded.stats
        _earned.value = decoded.badges
        _justAwarded.value = decoded.awarded.mapNotNull(BadgeKind::fromWire)
    }

    suspend fun refreshLeaderboard() = coroutineScope {
        val all = async { fetchBoard(null) }
        val month = async { fetchBoard("month") }
        _leaderboard.value = all.await()
        _leaderboardMonth.value = month.await()
    }

    private suspend fun fetchBoard(period: String?): List<LeaderRow> = runCatching {
        val url = "${Backend.WEB_API}/community/leaderboard" + (period?.let { "?period=$it" } ?: "")
        val resp = httpClient.get(url)
        if (resp.status.value != 200) emptyList()
        else lenientJson.decodeFromString<BoardPayload>(resp.bodyAsText()).rows
    }.getOrDefault(emptyList())

    @Serializable
    private data class MinePayload(val stats: ContributionStats? = null, val badges: List<EarnedBadge> = emptyList(), val awarded: List<String> = emptyList())

    @Serializable
    private data class BoardPayload(val rows: List<LeaderRow> = emptyList())
}

File: Farmsy/Core/Contributions.swift
import Foundation
import Observation

// Badges and the leaderboard: what a person gave the map, as the server counts
// it (migration 069). The app only reads; award_badges runs server-side after
// every report and post, so nothing here can be earned by tapping.

/// The eight badges, in display order, with their rule in words.
enum BadgeKind: String, CaseIterable, Identifiable {
    case firstReport = "first_report"
    case earlyBird = "early_bird"
    case confirmer
    case explorer
    case shelfScout = "shelf_scout"
    case photographer
    case regular
    case collector

    var id: String { rawValue }

    var title: String {
        switch self {
        case .firstReport: String(localized: "First report")
        case .earlyBird: String(localized: "Early bird")
        case .confirmer: String(localized: "Confirmer")
        case .explorer: String(localized: "Explorer")
        case .shelfScout: String(localized: "Shelf scout")
        case .photographer: String(localized: "Photographer")
        case .regular: String(localized: "Regular")
        case .collector: String(localized: "Collector")
        }
    }

    var rule: String {
        switch self {
        case .firstReport: String(localized: "Tell the app whether a farm was open, once.")
        case .earlyBird: String(localized: "Be the first to report a farm that day, five times.")
        case .confirmer: String(localized: "Confirm what someone else reported that day, ten times.")
        case .explorer: String(localized: "Report at ten different farms.")
        case .shelfScout: String(localized: "Say what was on the shelf in 25 reports.")
        case .photographer: String(localized: "Post a photo from a farm.")
        case .regular: String(localized: "Report in four different weeks.")
        case .collector: String(localized: "Save five farms.")
        }
    }

    var icon: String {
        switch self {
        case .firstReport: "flag.fill"
        case .earlyBird: "sunrise.fill"
        case .confirmer: "checkmark.seal.fill"
        case .explorer: "map.fill"
        case .shelfScout: "basket.fill"
        case .photographer: "camera.fill"
        case .regular: "calendar"
        case .collector: "heart.fill"
        }
    }
}

struct ContributionStats: Decodable, Sendable {
    let reports: Int
    let farms: Int
    let confirmations: Int
    let earlyBirds: Int
    let shelfReports: Int
    let weeks: Int
    let posts: Int
    let photoPosts: Int
    let saved: Int
    let badges: Int

    enum CodingKeys: String, CodingKey {
        case reports, farms, confirmations, weeks, posts, saved, badges
        case earlyBirds = "early_birds", shelfReports = "shelf_reports", photoPosts = "photo_posts"
    }
}

struct EarnedBadge: Decodable, Sendable {
    let badge: String
    let earnedAt: String
    enum CodingKeys: String, CodingKey { case badge, earnedAt = "earned_at" }
}

struct LeaderRow: Decodable, Identifiable, Sendable {
    let userId: String
    let name: String
    let reports: Int
    let farms: Int
    let confirmations: Int
    let badges: Int
    var id: String { userId }
    enum CodingKeys: String, CodingKey {
        case userId = "user_id", name, reports, farms, confirmations, badges
    }
}

@MainActor
@Observable
final class Contributions {
    static let shared = Contributions()
    private(set) var stats: ContributionStats?
    private(set) var earned: [EarnedBadge] = []
    /// Badges the last refresh awarded — the profile shows a celebration for
    /// these once, then forgets them.
    private(set) var justAwarded: [BadgeKind] = []
    private(set) var leaderboard: [LeaderRow] = []
    private(set) var leaderboardMonth: [LeaderRow] = []
    private init() {}

    func has(_ kind: BadgeKind) -> Bool { earned.contains { $0.badge == kind.rawValue } }
    func clearAwarded() { justAwarded = [] }

    func refreshMine(token: String) async {
        var request = URLRequest(url: Backend.webAPI.appending(path: "profile").appending(path: "contributions"))
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        guard let (data, resp) = try? await URLSession.shared.data(for: request),
              (resp as? HTTPURLResponse)?.statusCode == 200,
              let decoded = try? JSONDecoder().decode(MinePayload.self, from: data)
        else { return }
        stats = decoded.stats
        earned = decoded.badges
        justAwarded = decoded.awarded.compactMap(BadgeKind.init(rawValue:))
    }

    func refreshLeaderboard() async {
        async let all = fetchBoard(period: nil)
        async let month = fetchBoard(period: "month")
        leaderboard = await all
        leaderboardMonth = await month
    }

    private func fetchBoard(period: String?) async -> [LeaderRow] {
        var url = Backend.webAPI.appending(path: "community").appending(path: "leaderboard")
        if let period { url.append(queryItems: [URLQueryItem(name: "period", value: period)]) }
        guard let (data, resp) = try? await URLSession.shared.data(from: url),
              (resp as? HTTPURLResponse)?.statusCode == 200,
              let decoded = try? JSONDecoder().decode(BoardPayload.self, from: data)
        else { return [] }
        return decoded.rows
    }

    private struct MinePayload: Decodable { let stats: ContributionStats?; let badges: [EarnedBadge]; let awarded: [String] }
    private struct BoardPayload: Decodable { let rows: [LeaderRow] }
}
